atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBoundStore.java

```diff
@@ -80,117 +66,143 @@ private CassandraTimestampBoundStore(CassandraClientPool clientPool) {
+            DebugLogger.logger.info("[GET] The service is starting up. Attempting to get timestamp bound from the DB "
+                    + "and resetting it with this process's ID!");
+            TimestampBoundStoreEntry newEntry = TimestampBoundStoreEntry.create(entryInDb.getTimestamp(), id);
+            casWithRetry(entryInDb, newEntry);
+            startingUp = false;
+            return newEntry;
+        }
+        return entryInDb;
+    }
+
+    private void casWithRetry(TimestampBoundStoreEntry entryInDb, TimestampBoundStoreEntry newEntry) {
+        clientPool.runWithRetry((FunctionCheckedException<Client, Void, RuntimeException>) client -> {
+            cas(client, entryInDb, newEntry);
+            return null;
         });
     }
 
-    private void cas(Client client, Long oldVal, long newVal) {
-        final CASResult result;
+    private void cas(Client client, TimestampBoundStoreEntry entryInDb, TimestampBoundStoreEntry newEntry) {
+        checkLimitNotDecreasing(newEntry);
+        CASResult result = updateTimestampInDb(client, entryInDb, newEntry);
+        if (result.isSuccess()) {
+            setCurrentLimit("[CAS]", newEntry);
+        } else {
+            retryCasIfMatchingId(client, newEntry, result);
+        }
+    }
+
+    private void checkLimitNotDecreasing(TimestampBoundStoreEntry newEntry) {
+        if (currentLimit > newEntry.getTimestamp()) {
+            CassandraTimestampUtils.throwNewTimestampTooSmallException(currentLimit, newEntry);
+        }
+    }
+
+    private CASResult updateTimestampInDb(Client client, TimestampBoundStoreEntry entryInDb,
+            TimestampBoundStoreEntry newEntry) {
+        CASResult result;
+        DebugLogger.logger.info("[CAS] Trying to set upper limit from {} to {}.", entryInDb.getTimestampAsString(),
+                newEntry.getTimestampAsString());
         try {
-            DebugLogger.logger.info("[CAS] Trying to set upper limit from {} to {}.", oldVal, newVal);
             result = client.cas(
-                    getRowName(),
+                    BYTE_BUFFER_ROW_NAME,
                     AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName(),
-                    oldVal == null ? ImmutableList.of() : ImmutableList.of(makeColumn(oldVal)),
-                    ImmutableList.of(makeColumn(newVal)),
+                    makeListOfColumnsFromEntry(entryInDb),
+                    makeListOfColumnsFromEntry(newEntry),
                     ConsistencyLevel.SERIAL,
                     ConsistencyLevel.EACH_QUORUM);
+            return result;
         } catch (Exception e) {
-            log.error("[CAS] Error trying to set from {} to {}", oldVal, newVal, e);
-            DebugLogger.logger.error("[CAS] Error trying to set from {} to {}", oldVal, newVal, e);
-            throw Throwables.throwUncheckedException(e);
+            CassandraTimestampUtils.throwUpdateUncheckedException(entryInDb, newEntry, e);
+            return null;
         }
```

Comment:
Actually we can just use the method from Utils to log the exception, we don't actually need to pass it around